File: 78-subsets/subsets.cpp
class Solution {
public:
    void recur(vector<int>& nums, vector<vector<int>> &ans, vector<int> &temp, int i){
        int n = nums.size();
        if(i >= n){
            ans.push_back(temp);
            return;
        } else {
            temp.push_back(nums[i]);
            recur(nums, ans, temp, i+1);
            temp.pop_back();
            recur(nums, ans, temp, i+1);
        }
    }
    vector<vector<int>> subsets(vector<int>& nums) {
        vector<vector<int>> ans;
        vector<int> temp;
        recur(nums, ans, temp, 0);
        return ans;
    }
};

File: 78-subsets/subsets.py
class Solution:
    def recur(self, i, nums, temp, ans):
        if i >= len(nums):
            ans.append(temp.copy())
            return
        # Not take
        self.recur(i+1, nums, temp, ans)

        # Take
        temp.append(nums[i])
        self.recur(i+1, nums, temp, ans)
        temp.pop()
        # self.recur(i)
        # for j in range(i, len(nums)):
        #     # Take  
        #     temp.append(nums[j])
        #     self.recur(j+1, nums, temp, ans)
        #     temp.pop()

    def subsets(self, nums: List[int]) -> List[List[int]]:
        ans = []
        self.recur(0, nums, [], ans)
        return ans